### DISCUSSION TITLE
[SOLVED] Unable to import JSON feed

### DISCUSSION THREAD

**Discussion by mr-tchoo** [martinrotter/rssguard#875] [Q&A]
Hello! Whenever I tried to import a preprocessed json feed in a .json file such as:
`{
    "version": "https://jsonfeed.org/version/1",
    "title": "My Example Feed",
    "home_page_url": "https://example.org/",
    "feed_url": "https://example.org/feed.json",
    "items": [
        {
            "id": "2",
            "content_text": "This is a second item.",
            "url": "https://example.org/second-item"
        },
        {
            "id": "1",
            "content_html": "<p>Hello, world!</p>",
            "url": "https://example.org/initial-post"
        }
    ]
}`

or  simply 

`{"title": "example_title"}`

with following source script

`bash -c "less %data%/scripts/filename.json"`

and tried to fetch metadata, I got 'Error: XML is not well-formed, early end of document'.

I have also tried to use cat instead of less but that isnt working as well.

I would be glad to get some help :)

![image](https://user-images.githubusercontent.com/53568958/213523683-00bf7e2e-7ecc-4712-9d0b-1d90cd98531d.png)

**Comment by martinrotter** [comment]
Hi `less` is clearly a command you do not want to use because it is interactive text command line "pager".

`cat` is good and should clearly work, unless your "filename.json" file is not actually valid JSON feed file. Can you post "filename.json" sample file here for test?

**Reply by mr-tchoo** [reply]
Thanks for your quick response. I've tried `bash -c "cat %data%/scripts/filename.txt"` and got `Script failed: script threw an error: 'cat: 'C:/Users/Username/AppData/Local\RSS Guard 4/scripts/filename.txt': No such file or directory` although the exact file is located there.
Attached a sample file rename to .txt because .json is not allowed for file uploads seemingly.  
[filename.txt](https://github.com/martinrotter/rssguard/files/10460084/filename.txt)

**Reply by martinrotter** [reply]
OK, I will try tomorrow morning, now dont have time, thanks for input.

**Comment by akinokonomi** [comment]
@floatinger And if you try it like this?

```sh
bash -c "cat %data%\\scripts\\filename.json"
```
Or
```sh
bash -c 'cat "%data%\\scripts\\filename.json"'
```
Or
```sh
bash -c 'cat %data%"\\scripts\\filename.json"'
```

You might also try something like this?
```cmd
cmd.exe "type %data%\scripts\filename.json"
```

**Reply by mr-tchoo** [reply]
Hi @akinokonomi :)
First three give `Script failed: script threw an error: 'Error: 0xffffffff'`
Last one gives `Script failed: script execution took too long`
Though they all work typed into a regular terminal in the %data% directory.

**Reply by akinokonomi** [reply]
Ok. Just tried it myself. It doesn't work for me either...

This one works for me (Linux)
```bash
bash -c "cat "'%data%'"/filename.json"
```

This one too (more universal)
```bash
bash -c "cat "'"%data%"/filename.json'"
```

So, on Windows, following should do (from what I read the `\\` *must* be quoted):

```bash
bash -c "cat "'"%data%"\\filename.json'"
```

If not, try this:

```bash
bash -c "cat "%data%"'"\\filename.json'"
```

---

@martinrotter 

This does not look good. Here is the log output

```
core: Running custom script for guessing 'bash -c "cat "'"%data%"/filename.xml'"' to obtain feed data.
```

The command is encolsed in additional single quotes `'`.

It would be nice if RSS Guard could escape any other single quotes automatically `'"'"'` before executing the command.

But if you could, I think those additional single quotes might be dropped really.

**Reply by martinrotter** [reply]
Yes. I found the problem.

Problem is that RSS Guard mixed up directory path separators "/" and "\" on Windows. I will normalize normalise these to be native on each platform.

**Reply by martinrotter** [reply]
Possibly fixed with ebdde964.

Try devbuild and report back. Now it should use NATIVE path separators on each platform. So "/" on Unix-like OSes and "\\" on Windows. Be careful about quoting.

**Reply by martinrotter** [reply]
Test. No quoting seems to be needed. Also, as stated in documentation.

RSS Guard "tokenizes" your "script" input string to understand what is the executable ("powershell" in this case) and what are the actual arguments. It is highly recommended to use DOUBLE QUOTES to perform any kind of quoting or string ops.

![image](https://user-images.githubusercontent.com/1255302/214227256-b21cd63c-6efe-4aa6-95fa-798e72b5dfab.png)

**Reply by akinokonomi** [reply]
> No quoting seems to be needed. Also, as stated in documentation.

Yeah, but it's only do good for Linux and Mac, because PATH works as it should there.

You can just use `cat path/to/file`. Not on Windows though.

On windows, to use bash, cmd.exe, quoting is needed often enough (I think).

> RSS Guard "tokenizes" your "script" input string to understand what is the executable ("powershell" in this case)

Hm. I think it's just because powershell is what `path/to/powershell.exe` named/linked/etc. in Windows PATH. No?

> highly recommended to use DOUBLE QUOTES

Because RSS Guard encloses the whole command in an additional single quotes?

I've had enough problems with additional, invisible single quotes, I'd really recommend to somehow get rid of that here.

**Reply by akinokonomi** [reply]
@martinrotter Just tested 2e76ae5b :
```bash
bash -c "cat %data%/filename.json" # does not work
bash -c 'cat %data%/filename.json' # does not work
bash -c "cat '"%data%"'/filename.json" # works
```

This is how the third command seem to be interpreted:
```bash
'bash -c "cat '"%data%"'/filename.json"'
```

**Reply by martinrotter** [reply]
> Because RSS Guard encloses the whole command in an additional single quotes?
No, it does not.

Like I wrote. Should be fixed now. The issue was really Windows-specific, on Linux, no changes. No quoting for backslashes is needed on Windows as demonstrated in screenshot.

**Reply by akinokonomi** [reply]
> The issue was really Windows-specific, on Linux, no changes.

So it was about backslashes only. Got it.

**Reply by martinrotter** [reply]
Effectively, yes. On all other platforms Qt uses forward slashes internally too, so no problem. On  Windows, things sometimes mix up.

**Reply by martinrotter** [reply]
Also, issue could be workarounded by some simple bash magic.

`bash -c "echo 'C:/data\world' | tr '/' '\'"`

**Reply by martinrotter** [reply]
So for example

`bash -c "cat $(echo -n '%data%' | tr '\' '/')/scripts/filename.json"`

But I guess that Bash on Windows does not use "C:...." as drives and rather it uses "/c/....". So to have like fully portable script callings, it is better to not use these oneliners and write bash script instead and call the script and do path conversions inside script for better readablitity.

**Reply by mr-tchoo** [reply]
`powershell "cat %data%\scripts\filename.json"` actually does not work with my non-portable data folder configuration on Windows because the **"RSS Guard 4"** folder name is seperated by a whitespace, so it throws `[...] No positional argument found that accepts argument "Guard" [...]`.
`powershell "cat '%data%\scripts\filename.json'"` does work though. I'm on RSSGuard 4.3.1.
